File: Week05/LeetCode-0496-2.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        """
        O(m+n) algorithm with O(n) extra space
        """
        S = []
        myMap = dict()
        for num in nums2[-1::-1]:
            while len(S) != 0 and num > S[-1]:
                S.pop()
             
            if len(S) == 0:
                myMap[num] = -1
            else:
                myMap[num] = S[-1]
            S.append(num)
        
        result = []
        for num in nums1:
            result.append(myMap[num])
        return result

File: Week05/LeetCode-0496-2.cpp
class Solution {
public:
    ///-------------------------------------------------------------------------
    /// This is an O(n+m) algorithm that uses O(m) extra space
    ///
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> nextGreater;
        stack<int> S;

        for (int i=(int)nums2.size()-1; i>=0; i--){
            while (!S.empty() && nums2[i] >= S.top()) S.pop();

            if (S.empty()) nextGreater[nums2[i]] = -1;
            else           nextGreater[nums2[i]] = S.top();

            S.push(nums2[i]);
        } //end-for

        // Find nums1[i] in the map & push the next greater to the result
        vector<int> result;
        for (int i=0; i<nums1.size(); i++){
            result.push_back(nextGreater[nums1[i]]);
        } //end-for

        return result;
    } //end-nextGreaterElement
};